### PR TITLE
Add ruff

### PR DESCRIPTION
**Pull request by jo47011**
Replace all the original project setup, linting, etc. with ruff.

Taking it step by step.  We could have a PR for each step to simplify code reading and discussions.

1. **Ruff:** Introduce ruff.  
    We disable all linting errors that we get so we can focus on the project setup 1st.  
    
2. **Formatting:** We agree on some parameters, see `[tool.ruff.format]` in `pyproject.toml`.  Then we reformat the code (replaces `black`):
    ```bash
    ruff format .
    ````
    Then we can also add a format check to the build workflow.
    
3. **Linting:** We agree on which parameter we want to enable and which we want to ignore in future, see `[tool.ruff.lint]` in `pyproject.toml`, e.g.;
```toml
    "D106",  # Missing docstring in public nested class
    "D107",  # Ignore D107 Missing docstring  # FIXME: enable again???
    "D200",  # One-line docstring should fit on one line
    ...
```
invoke like this:
```bash
ruff check --fix .
````
4. **Build-system:** we could consider switching from `Setuptools` to `Hatchling` further down the road.  Provides for some more advanced environment management and other nice features.
 
BTW I think there was a bug in your previous setup.cfg:
image

I fixed it in the `pyproject.toml`.  If it was intentional we need to change it back.

We should also discuss the skipping of the tests in `test_midi.py`.   There where skipped because `python-rtmidi` was not installed but `midi` was.  Does this make sense?  I changed the logic slightly and now catch the exception as well if module `midi` is not loaded.  See my comments in the code for details.

### CODE REVIEW

**Comment by jo47011** [comment]
Sorry this one was to early.  Wanted a PR in my fork only so you do not get all the notifications.  Will close this one and open a new one once I'm ready.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -103,7 +100,9 @@ jobs:
     - name: Install Python dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install -r requirements.txt
+        mkdir _static
```
Comment (jo47011) [inline .github/workflows/build.yml:103]: Do we want this here to avoid the warning?

`WARNING: html_static_path entry '_static' does not exist`

File: pyproject.toml [hunk 2]
```diff
@@ -1,3 +1,136 @@
...
+    "venv",
+    "venv*",
+    ".mypy_cache",
+    "test/assets",
+]
+
+[tool.ruff.lint]
+# Enable checks for Pycodestyle, Pyflakes, Bugbear, Docstrings and Import sorting
+select = ["E", "F", "B", "D", "I"]
+ignore = [
```
Comment (jo47011) [inline pyproject.toml:66]: All the errors below occur.  As proposed we can enable and fix them one by one in step 3.

File: scripts/extract_dependencies.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+"""Extract the names int the [project.optional-dependencies] section of a pyproject.toml file.
```
Comment (jo47011) [inline scripts/extract_dependencies.py:1]: Not sure where to put the script.  Let me know whether you prefer it somewhere else.  

I don't like the need for such a script in the 1st place but there seems to be no generic way like `pip install .[all]` when using `pyproject.toml`.  Using this scripts avoids repeating dependencies which would be hard to maintain in the long run, e.g.
```python
   pip install .[mysql,knx,dmx,midi,mqtt,pulse,telegram,file_persistence]
```
So for me the script is the better choice.

File: test/interfaces/test_midi.py [hunk 4]
```diff
@@ -2,14 +2,14 @@
...
-from shc.datatypes import RangeUInt8
-
 from .._helper import InterfaceThreadRunner
 
 
 try:
+    import mido
+    import shc.interfaces.midi
+    from shc.datatypes import RangeUInt8
+
```
Comment (jo47011) [inline test/interfaces/test_midi.py:12]: Needed to move the import here.

Avoids the `pip uninstall python-rtmidi` to skip those tests.  Not sure whether this is really the best logic as this happens quietly and might not be noticed if skipped accidently.

File: test/interfaces/test_midi.py [hunk 5]
```diff
@@ -18,6 +18,7 @@
     mido_backend_error = str(e)
 
 
+@unittest.skipUnless(mido_backend_available, "mido MIDI backend is not available: {}".format(mido_backend_error))
```
Comment (jo47011) [inline test/interfaces/test_midi.py:21]: This line is new to skip this test as well.  Other `@unittest.skipUnless` changes is a typo fix only ("available").

File: test/test_interfaces_stub.py [hunk 6]
```diff
@@ -7,6 +7,3 @@ def test_knx(self):
 
     def test_dmx(self):
         import shc.interfaces.dmx  # noqa: F401
-
-    def test_midi(self):
-        import shc.interfaces.midi  # noqa: F401
```
Comment (jo47011) [inline test/test_interfaces_stub.py:12]: Had to remove this one.  Well it is tested in test_midi.py which is skipped in the pipeline.  

Is my current proposal sufficient or do you feel like we need this test here?  Then we rollback to your approach and install `midi` but uninstall `python-rtmidi` which is kind of confusing to me.  Maybe we should split the dependency then.

**Comment by jo47011** [comment]
As for the failed code coverage report, I suggest we fix it once we have the test strategy for the `midi` stuff clarified.
Can't we just mock the failing midi tests:
```sh
_rtmidi.SystemError: MidiInAlsa::initialize: error creating ALSA sequencer client object.
```

**Comment by jo47011** [comment]
Hi @mhthies, don't want to push you (seriously not).  Just want to make sure we're on the same page.  I would like you to make a 1st review and provide some feedback.  I know not all checks are passing right now but the fix would depend on the result of our discussion.  Feel assigned as reviewer to this PR 😉 No rush, I'm also quite busy right now.

File: setup.py
Comment (mhthies) [inline setup.py:1]: I think the `setup.py` is still required, even with all the other changes. Otherwise, the web assets are not installed and compiled to be packaged in the wheel distribution. Or did I overlook something?

**Comment by mhthies** [comment]
On the midi tests:

In general, the MIDO tests only work when one of the MIDO backends is installed. Thus, I decided to make skipping the tests dependent on the availability of a MIDO backend.

In addition, we have the problem that the MIDI tests fail on GitHub Actions due to a missing kernel module/option which is required for creating MIDI ports, including loopback ports (i.e. MIDO output ports with `virtual=True`). Uninstalling the `rt-midi` package on GitHub Actions is a workaround to get the correct behavior: Although, *technically* a MIDO backend is available on Linux, using rt-midi, it is not functional in GitHub Actions. So, the tests should treat this platform as if there were no MIDO backend.

To get at least some basic coverage of the Python code, we have the "stub tests" which simply imports the python module without executing any function. This would still detect syntax or import errors.

Do you see a reason to change this behavior?

**Comment by jo47011** [comment]
> In general, the MIDO tests only work when one of the MIDO backends is installed. Thus, I decided to make skipping the tests dependent on the availability of a MIDO backend.
> 
> In addition, we have the problem that the MIDI tests fail on GitHub Actions due to a missing kernel module/option which is required for creating MIDI ports, including loopback ports (i.e. MIDO output ports with `virtual=True`). Uninstalling the `rt-midi` package on GitHub Actions is a workaround to get the correct behavior: Although, _technically_ a MIDO backend is available on Linux, using rt-midi, it is not functional in GitHub Actions. So, the tests should treat this platform as if there were no MIDO backend.

I understand your approach and the missing option of having those tests run on GitHub Actions.  I just didn't like "pip uninstall python-rtmidi" in the build.yml to achieve it.  But I have no strong opinion on that so I will roll back the changes to keep the diffs of this PR as small as possible.

> 
> To get at least some basic coverage of the Python code, we have the "stub tests" which simply imports the python module without executing any function. This would still detect syntax or import errors.
> 
> Do you see a reason to change this behavior?

To be honest I did not see the `detect syntax or import errors` benefit.  As for this reason I now understand your approach.

File: setup.py
Comment (jo47011) [reply setup.py:1]: You are totally right.  I missed out on the npm install.  I'll put it back in.

File: scripts/extract_dependencies.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    if exclude is None:
+        exclude = []
+
+    with open(toml_file_path, "r") as f:
+        toml_content = f.read()
+
+    # Find the optional-dependencies section
+    match = re.search(r"\[project.optional-dependencies\](.*?)\n\n", toml_content, re.DOTALL)
+    if not match:
+        return []
```
Comment (mhthies) [inline scripts/extract_dependencies.py:25]: Please, don't parse TOML with a RegEx.

I see that we cannot reliably use `tomli` as a dependency for Python  < 3.10 here, cause this script is used to to figure out the dependencies, in the first place. Maybe we should try to avoid this script alltogether and instead write the full list of extras/features into every pip invocation? I'd actually find that okay.

Yes, I know, I was okay with the separate list of dependencies in the requirements.txt, as well, which had obvious disadvantages. However, with a duplicated list of extras which does not match completely, we would more likely see a failure in the CI pipeline.

So, in conclusion, I'd prefer not using this script but hardcoding the list of extras/features.

**Review by mhthies — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 8]
```diff
@@ -1,3 +1,137 @@
...
+plugins = ["shc/util/mypy_variable_plugin.py"]
+exclude = "^test/assets/mypy_plugin_test"
+
+[[tool.mypy.overrides]]
+module = "mido,astropy,astropy.*,serial,serial_asyncio,paho.mqtt.*,selenium.*,pulsectl.*,pulsectl_asyncio,aiomysql,aiogram.*"
+ignore_missing_imports = true
+
+[tool.ruff.format]
+# Like Black, use double quotes for strings.
+# quote-style = "double"
```
[the remark this answers is not in the document]
Comment (jo47011) [reply pyproject.toml:127]: This would cause a lot of changes and the key changes of this PR could easily be overseen.  Once we have the strategy clear we do it step by step, e.g. one PR only for replacing ' => ".

File: README.md [hunk 9]
```diff
@@ -247,7 +258,20 @@ coverage html
 # open htmlcov/index.html
 ``` 
 
-We also enforce static type correctness with MyPy and Python codestyle rules with flake8.
-To run the static type checks and codestyle checks locally, simply install MyPy and flake8 and execute the `mypy` and `flake8` commands in the shc project repository.
+We also enforce static type correctness with MyPy and Python codestyle rules with ruff.
+To run the static type checks and codestyle checks locally, simply install MyPy and flake8 and execute the `mypy` command in the shc project repository.
```
[the remark this answers is not in the document]
Comment (jo47011) [reply README.md:262]: Remove sphinx config option.  Of course this is the better approach:
# html_static_path = ['_static']

Removed flak8 from README.

File: setup.cfg [hunk 10]
```diff
@@ -1,55 +0,0 @@
...
-midi =
-    mido>=1.2.9,<2
-    python-rtmidi>=1.4.6,<2
-mqtt = aiomqtt>=2.0.0,<3
-pulse = pulsectl_asyncio>=1.0.0,<2
-telegram = aiogram>=3.0,<4
-file_persistence = aiofile>=3.7.4,<4
-
-[options.package_data]
-shc.web = static/pack/*
```
[the remark this answers is not in the document]
Comment (jo47011) [reply setup.cfg:50]:
```bash
python -m build --sdist

tar tvzf ./dist/smarthomeconnect-0.9.0.tar.gz

smarthomeconnect-0.9.0/
smarthomeconnect-0.9.0/LICENSE
smarthomeconnect-0.9.0/MANIFEST.in
smarthomeconnect-0.9.0/NOTICE
smarthomeconnect-0.9.0/PKG-INFO
smarthomeconnect-0.9.0/README.md
smarthomeconnect-0.9.0/package.json
smarthomeconnect-0.9.0/pyproject.toml
smarthomeconnect-0.9.0/setup.cfg
smarthomeconnect-0.9.0/setup.py
smarthomeconnect-0.9.0/shc/
smarthomeconnect-0.9.0/shc/__init__.py
smarthomeconnect-0.9.0/shc/base.py
smarthomeconnect-0.9.0/shc/conversion.py
smarthomeconnect-0.9.0/shc/data_logging.py
smarthomeconnect-0.9.0/shc/datatypes.py
smarthomeconnect-0.9.0/shc/expressions.py
smarthomeconnect-0.9.0/shc/interfaces/
smarthomeconnect-0.9.0/shc/interfaces/__init__.py
smarthomeconnect-0.9.0/shc/interfaces/_helper.py
smarthomeconnect-0.9.0/shc/interfaces/_pulse_ffi.py
smarthomeconnect-0.9.0/shc/interfaces/command.py
smarthomeconnect-0.9.0/shc/interfaces/dmx.py
smarthomeconnect-0.9.0/shc/interfaces/file_persistence.py
smarthomeconnect-0.9.0/shc/interfaces/in_memory_data_logging.py
smarthomeconnect-0.9.0/shc/interfaces/knx.py
smarthomeconnect-0.9.0/shc/interfaces/midi.py
smarthomeconnect-0.9.0/shc/interfaces/mqtt.py
smarthomeconnect-0.9.0/shc/interfaces/mysql.py
smarthomeconnect-0.9.0/shc/interfaces/ping.py
smarthomeconnect-0.9.0/shc/interfaces/pulse.py
smarthomeconnect-0.9.0/shc/interfaces/shc_client.py
smarthomeconnect-0.9.0/shc/interfaces/system_monitoring.py
smarthomeconnect-0.9.0/shc/interfaces/tasmota.py
smarthomeconnect-0.9.0/shc/interfaces/telegram.py
smarthomeconnect-0.9.0/shc/misc.py
smarthomeconnect-0.9.0/shc/py.typed
smarthomeconnect-0.9.0/shc/supervisor.py
smarthomeconnect-0.9.0/shc/timer.py
smarthomeconnect-0.9.0/shc/util/
smarthomeconnect-0.9.0/shc/util/__init__.py
smarthomeconnect-0.9.0/shc/util/check_shc.py
smarthomeconnect-0.9.0/shc/util/mypy_variable_plugin.py
smarthomeconnect-0.9.0/shc/variables.py
smarthomeconnect-0.9.0/shc/web/
smarthomeconnect-0.9.0/shc/web/__init__.py
smarthomeconnect-0.9.0/shc/web/interface.py
smarthomeconnect-0.9.0/shc/web/log_widgets.py
smarthomeconnect-0.9.0/shc/web/static/
smarthomeconnect-0.9.0/shc/web/static/apple-touch-icon.png
smarthomeconnect-0.9.0/shc/web/static/favicon.ico
smarthomeconnect-0.9.0/shc/web/templates/
smarthomeconnect-0.9.0/shc/web/templates/base.htm
smarthomeconnect-0.9.0/shc/web/templates/buttons.inc.htm
smarthomeconnect-0.9.0/shc/web/templates/log/
smarthomeconnect-0.9.0/shc/web/templates/log/chart.htm
smarthomeconnect-0.9.0/shc/web/templates/log/loglist.htm
smarthomeconnect-0.9.0/shc/web/templates/page.htm
smarthomeconnect-0.9.0/shc/web/templates/status.htm
smarthomeconnect-0.9.0/shc/web/templates/widgets/
smarthomeconnect-0.9.0/shc/web/templates/widgets/buttongroup.htm
smarthomeconnect-0.9.0/shc/web/templates/widgets/colorchoser.htm
smarthomeconnect-0.9.0/shc/web/templates/widgets/hiderowbox.htm
smarthomeconnect-0.9.0/shc/web/templates/widgets/imagemap.htm
smarthomeconnect-0.9.0/shc/web/templates/widgets/select.htm
smarthomeconnect-0.9.0/shc/web/templates/widgets/slider.htm
smarthomeconnect-0.9.0/shc/web/templates/widgets/switch.htm
smarthomeconnect-0.9.0/shc/web/templates/widgets/textdisplay.htm
smarthomeconnect-0.9.0/shc/web/templates/widgets/textinput.htm
smarthomeconnect-0.9.0/shc/web/widgets.py
smarthomeconnect-0.9.0/smarthomeconnect.egg-info/
smarthomeconnect-0.9.0/smarthomeconnect.egg-info/PKG-INFO
smarthomeconnect-0.9.0/smarthomeconnect.egg-info/SOURCES.txt
smarthomeconnect-0.9.0/smarthomeconnect.egg-info/dependency_links.txt
smarthomeconnect-0.9.0/smarthomeconnect.egg-info/requires.txt
smarthomeconnect-0.9.0/smarthomeconnect.egg-info/top_level.txt
smarthomeconnect-0.9.0/test/
smarthomeconnect-0.9.0/test/test_base.py
smarthomeconnect-0.9.0/test/test_conversion.py
smarthomeconnect-0.9.0/test/test_data_logging.py
smarthomeconnect-0.9.0/test/test_datatypes.py
smarthomeconnect-0.9.0/test/test_examples.py
smarthomeconnect-0.9.0/test/test_expressions.py
smarthomeconnect-0.9.0/test/test_interfaces_stub.py
smarthomeconnect-0.9.0/test/test_misc.py
smarthomeconnect-0.9.0/test/test_timer.py
smarthomeconnect-0.9.0/test/test_variables.py
smarthomeconnect-0.9.0/test/test_web.py
smarthomeconnect-0.9.0/web_ui_src/
smarthomeconnect-0.9.0/web_ui_src/main.css
smarthomeconnect-0.9.0/web_ui_src/main.js
smarthomeconnect-0.9.0/web_ui_src/prism.png
smarthomeconnect-0.9.0/web_ui_src/shc_base.js
smarthomeconnect-0.9.0/web_ui_src/widgets/
smarthomeconnect-0.9.0/web_ui_src/widgets/basic_widgets.js
smarthomeconnect-0.9.0/web_ui_src/widgets/colorchoser_widget.css
smarthomeconnect-0.9.0/web_ui_src/widgets/colorchoser_widget.js
smarthomeconnect-0.9.0/web_ui_src/widgets/hiderow_widget.css
smarthomeconnect-0.9.0/web_ui_src/widgets/log_list_widget.css
smarthomeconnect-0.9.0/web_ui_src/widgets/log_widgets.js
smarthomeconnect-0.9.0/web_ui_src/widgets/slider_widget.css
smarthomeconnect-0.9.0/web_ui_src/widgets/switch_widget.css
```

File: setup.cfg [hunk 11]
```diff
@@ -1,55 +0,0 @@
...
-
-[options.packages.find]
-exclude=
-    test
-    test.*
```
[the remark this answers is not in the document]
Comment (jo47011) [reply setup.cfg:55]:
```bash
python -m build --wheel

unzip -l dist/smarthomeconnect-0.9.0-py3-none-any.whl

Archive:  dist/smarthomeconnect-0.9.0-py3-none-any.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
      835  2024-10-19 13:20   shc/__init__.py
    30795  2024-12-04 10:51   shc/base.py
     3642  2024-12-04 10:51   shc/conversion.py
    36924  2024-12-04 10:51   shc/data_logging.py
    12113  2024-12-04 10:51   shc/datatypes.py
    29107  2024-12-04 10:51   shc/expressions.py
    20590  2024-12-04 10:51   shc/misc.py
        0  2024-10-19 13:20   shc/py.typed
     9588  2024-12-04 10:51   shc/supervisor.py
    54705  2024-12-04 10:51   shc/timer.py
    10452  2024-12-04 10:51   shc/variables.py
        0  2024-10-19 13:20   shc/interfaces/__init__.py
    20544  2024-12-04 10:51   shc/interfaces/_helper.py
     2345  2024-12-04 10:51   shc/interfaces/_pulse_ffi.py
     4829  2024-12-04 10:51   shc/interfaces/command.py
     8417  2024-12-04 10:51   shc/interfaces/dmx.py
    10022  2024-12-04 10:51   shc/interfaces/file_persistence.py
     3755  2024-12-04 10:51   shc/interfaces/in_memory_data_logging.py
    17325  2024-12-04 10:51   shc/interfaces/knx.py
    19371  2024-12-04 10:51   shc/interfaces/midi.py
    18925  2024-12-04 10:51   shc/interfaces/mqtt.py
    16625  2024-12-04 10:51   shc/interfaces/mysql.py
     2963  2024-12-04 10:51   shc/interfaces/ping.py
    37059  2024-12-04 10:51   shc/interfaces/pulse.py
    19493  2024-12-04 10:51   shc/interfaces/shc_client.py
     5569  2024-12-04 10:51   shc/interfaces/system_monitoring.py
    37117  2024-12-04 10:51   shc/interfaces/tasmota.py
    30646  2024-12-04 10:51   shc/interfaces/telegram.py
        0  2024-10-19 13:20   shc/util/__init__.py
     4463  2024-12-04 10:51   shc/util/check_shc.py
     3511  2024-12-04 10:51   shc/util/mypy_variable_plugin.py
       34  2024-12-04 10:51   shc/web/__init__.py
    56395  2024-12-04 10:51   shc/web/interface.py
    13620  2024-12-04 10:51   shc/web/log_widgets.py
    41909  2024-12-04 10:51   shc/web/widgets.py
     4727  2024-10-19 13:20   shc/web/static/apple-touch-icon.png
     1150  2024-10-19 13:20   shc/web/static/favicon.ico
   291164  2024-11-02 19:33   shc/web/static/pack/Lato-Bold.367f6330.woff
   178868  2024-11-04 21:47   shc/web/static/pack/Lato-Bold.59265338.woff2
   178868  2024-11-02 19:33   shc/web/static/pack/Lato-Bold.b0c7a5bb.woff2
   291164  2024-11-04 21:47   shc/web/static/pack/Lato-Bold.be904d71.woff
   188332  2024-11-02 19:33   shc/web/static/pack/Lato-BoldItalic.86b3f184.woff2
   303376  2024-11-02 19:33   shc/web/static/pack/Lato-BoldItalic.a844a9b2.woff
   303376  2024-11-04 21:47   shc/web/static/pack/Lato-BoldItalic.b5212685.woff
   188332  2024-11-04 21:47   shc/web/static/pack/Lato-BoldItalic.e844b943.woff2
   308540  2024-11-04 21:47   shc/web/static/pack/Lato-Italic.1c3dc9f6.woff
   190608  2024-11-02 19:33   shc/web/static/pack/Lato-Italic.c1782172.woff2
   308540  2024-11-02 19:33   shc/web/static/pack/Lato-Italic.eab68376.woff
   190608  2024-11-04 21:47   shc/web/static/pack/Lato-Italic.ef1b93fb.woff2
   178044  2024-11-04 21:47   shc/web/static/pack/Lato-Regular.1380e58c.woff2
   290956  2024-11-04 21:47   shc/web/static/pack/Lato-Regular.175e0a67.woff
   290956  2024-11-02 19:33   shc/web/static/pack/Lato-Regular.96b82b78.woff
   178044  2024-11-02 19:33   shc/web/static/pack/Lato-Regular.ebc7de56.woff2
    33616  2024-11-02 19:33   shc/web/static/pack/LatoLatin-Bold.8431ae36.woff2
    33616  2024-11-04 21:47   shc/web/static/pack/LatoLatin-Bold.b879207a.woff2
    45416  2024-11-02 19:33   shc/web/static/pack/LatoLatin-Bold.c7218d37.woff
    45416  2024-11-04 21:47   shc/web/static/pack/LatoLatin-Bold.f85d624b.woff
    46468  2024-11-02 19:33   shc/web/static/pack/LatoLatin-BoldItalic.26f6ffbc.woff
    46468  2024-11-04 21:47   shc/web/static/pack/LatoLatin-BoldItalic.70f3e341.woff
    34752  2024-11-04 21:47   shc/web/static/pack/LatoLatin-BoldItalic.a17079a7.woff2
    34752  2024-11-02 19:33   shc/web/static/pack/LatoLatin-BoldItalic.a8e39bbd.woff2
    34848  2024-11-04 21:47   shc/web/static/pack/LatoLatin-Italic.414fbd89.woff2
    34848  2024-11-02 19:33   shc/web/static/pack/LatoLatin-Italic.d2089e8b.woff2
    46980  2024-11-02 19:33   shc/web/static/pack/LatoLatin-Italic.de0574c8.woff
    46980  2024-11-04 21:47   shc/web/static/pack/LatoLatin-Italic.e333a4a5.woff
    33828  2024-11-04 21:47   shc/web/static/pack/LatoLatin-Regular.45b5be36.woff2
    45632  2024-11-04 21:47   shc/web/static/pack/LatoLatin-Regular.6f71ea1b.woff
    45632  2024-11-02 19:33   shc/web/static/pack/LatoLatin-Regular.a124b19d.woff
    33828  2024-11-02 19:33   shc/web/static/pack/LatoLatin-Regular.b321314c.woff2
    76736  2024-11-02 19:33   shc/web/static/pack/brand-icons.0b52c614.woff2
    76736  2024-11-04 21:47   shc/web/static/pack/brand-icons.8b67633d.woff2
    89988  2024-11-02 19:33   shc/web/static/pack/brand-icons.a0b02584.woff
    89988  2024-11-04 21:47   shc/web/static/pack/brand-icons.b7e862ad.woff
    78268  2024-11-02 19:33   shc/web/static/pack/icons.6002b354.woff2
   101648  2024-11-02 19:33   shc/web/static/pack/icons.80005ece.woff
   101648  2024-11-04 21:47   shc/web/static/pack/icons.c430880e.woff
    78268  2024-11-04 21:47   shc/web/static/pack/icons.c8ed78b5.woff2
  1602056  2025-01-05 20:36   shc/web/static/pack/main.css
  4046141  2025-01-05 20:36   shc/web/static/pack/main.css.map
  2119903  2024-11-04 21:47   shc/web/static/pack/main.js
  3762643  2024-11-04 21:47   shc/web/static/pack/main.js.map
    16276  2024-11-02 19:33   shc/web/static/pack/outline-icons.09c79cc5.woff
    16276  2024-11-04 21:47   shc/web/static/pack/outline-icons.2c14f812.woff
    13224  2024-11-02 19:33   shc/web/static/pack/outline-icons.5bedcec6.woff2
    13224  2024-11-04 21:47   shc/web/static/pack/outline-icons.e37142a3.woff2
     9697  2024-11-04 21:47   shc/web/static/pack/prism.403685ff.png
     8886  2024-11-02 19:33   shc/web/static/pack/prism.810162de.png
     4800  2024-12-01 15:28   shc/web/templates/base.htm
     1209  2024-10-19 13:20   shc/web/templates/buttons.inc.htm
     1558  2024-10-19 13:20   shc/web/templates/page.htm
     2685  2024-10-19 13:20   shc/web/templates/status.htm
      159  2024-10-19 13:20   shc/web/templates/log/chart.htm
      237  2024-10-19 13:20   shc/web/templates/log/loglist.htm
      423  2024-10-31 09:16   shc/web/templates/widgets/buttongroup.htm
       56  2024-10-19 13:20   shc/web/templates/widgets/colorchoser.htm
      509  2024-10-19 13:20   shc/web/templates/widgets/hiderowbox.htm
     1485  2024-10-19 13:20   shc/web/templates/widgets/imagemap.htm
      415  2024-10-19 13:20   shc/web/templates/widgets/select.htm
      587  2024-10-19 13:20   shc/web/templates/widgets/slider.htm
      430  2024-10-19 13:20   shc/web/templates/widgets/switch.htm
      169  2024-10-19 13:20   shc/web/templates/widgets/textdisplay.htm
      629  2024-10-19 13:20   shc/web/templates/widgets/textinput.htm
    11358  2025-01-16 08:03   smarthomeconnect-0.9.0.dist-info/LICENSE
    26964  2025-01-16 08:03   smarthomeconnect-0.9.0.dist-info/METADATA
      375  2025-01-16 08:03   smarthomeconnect-0.9.0.dist-info/NOTICE
       91  2025-01-16 08:03   smarthomeconnect-0.9.0.dist-info/WHEEL
        4  2025-01-16 08:03   smarthomeconnect-0.9.0.dist-info/top_level.txt
    10044  2025-01-16 08:03   smarthomeconnect-0.9.0.dist-info/RECORD
---------                     -------
 17458214                     108 files

```

File: pyproject.toml [hunk 12]
```diff
@@ -1,3 +1,137 @@
...
+    ".git",
+    "dist",
+    "docs",
+    "htmlcov",
+    "node_modules",
+    "venv",
+    "venv*",
+    ".mypy_cache",
+    "test/assets",
+]
```
[the remark this answers is not in the document]
Comment (jo47011) [reply pyproject.toml:62]: You are so right, changed it to:

```bash
include = ["shc/**/*.py", "test/**/*.py"]
exclude = ["test/assets"]
```
```bash
ruff check . --show-files
smarthomeconnect/shc/__init__.py
smarthomeconnect/shc/base.py
smarthomeconnect/shc/conversion.py
smarthomeconnect/shc/data_logging.py
smarthomeconnect/shc/datatypes.py
smarthomeconnect/shc/expressions.py
smarthomeconnect/shc/interfaces/__init__.py
smarthomeconnect/shc/interfaces/_helper.py
smarthomeconnect/shc/interfaces/_pulse_ffi.py
smarthomeconnect/shc/interfaces/command.py
smarthomeconnect/shc/interfaces/dmx.py
smarthomeconnect/shc/interfaces/file_persistence.py
smarthomeconnect/shc/interfaces/in_memory_data_logging.py
smarthomeconnect/shc/interfaces/knx.py
smarthomeconnect/shc/interfaces/midi.py
smarthomeconnect/shc/interfaces/mqtt.py
smarthomeconnect/shc/interfaces/mysql.py
smarthomeconnect/shc/interfaces/ping.py
smarthomeconnect/shc/interfaces/pulse.py
smarthomeconnect/shc/interfaces/shc_client.py
smarthomeconnect/shc/interfaces/system_monitoring.py
smarthomeconnect/shc/interfaces/tasmota.py
smarthomeconnect/shc/interfaces/telegram.py
smarthomeconnect/shc/misc.py
smarthomeconnect/shc/supervisor.py
smarthomeconnect/shc/timer.py
smarthomeconnect/shc/util/__init__.py
smarthomeconnect/shc/util/check_shc.py
smarthomeconnect/shc/util/mypy_variable_plugin.py
smarthomeconnect/shc/variables.py
smarthomeconnect/shc/web/__init__.py
smarthomeconnect/shc/web/interface.py
smarthomeconnect/shc/web/log_widgets.py
smarthomeconnect/shc/web/widgets.py
smarthomeconnect/test/__init__.py
smarthomeconnect/test/_helper.py
smarthomeconnect/test/interfaces/__init__.py
smarthomeconnect/test/interfaces/test_command.py
smarthomeconnect/test/interfaces/test_file_persistence.py
smarthomeconnect/test/interfaces/test_in_memory_data_logging.py
smarthomeconnect/test/interfaces/test_knx.py
smarthomeconnect/test/interfaces/test_midi.py
smarthomeconnect/test/interfaces/test_mqtt.py
smarthomeconnect/test/interfaces/test_mysql.py
smarthomeconnect/test/interfaces/test_ping.py
smarthomeconnect/test/interfaces/test_pulse.py
smarthomeconnect/test/interfaces/test_shc_client.py
smarthomeconnect/test/interfaces/test_system_monitoring.py
smarthomeconnect/test/interfaces/test_tasmota.py
smarthomeconnect/test/interfaces/test_telegram.py
smarthomeconnect/test/test_base.py
smarthomeconnect/test/test_conversion.py
smarthomeconnect/test/test_data_logging.py
smarthomeconnect/test/test_datatypes.py
smarthomeconnect/test/test_examples.py
smarthomeconnect/test/test_expressions.py
smarthomeconnect/test/test_interfaces_stub.py
smarthomeconnect/test/test_misc.py
smarthomeconnect/test/test_timer.py
smarthomeconnect/test/test_variables.py
smarthomeconnect/test/test_web.py
```

File: setup.cfg [hunk 13]
```diff
@@ -1,55 +0,0 @@
...
-classifiers =
-    Development Status :: 4 - Beta
-    Programming Language :: Python :: 3.8
-    Programming Language :: Python :: 3.9
-    Programming Language :: Python :: 3.10
-    Programming Language :: Python :: 3.11
-    Programming Language :: Python :: 3.12
-    License :: OSI Approved :: Apache Software License
-    Topic :: Home Automation
-    Framework :: AsyncIO
```
[the remark this answers is not in the document]
Comment (jo47011) [reply setup.cfg:25]: Whoops, good catch.  Added them to `pyproject.toml`:
```bash
classifiers = [
    "Programming Language :: Python :: 3.12",
    "License :: OSI Approved :: Apache Software License",
    "Topic :: Home Automation",
    "Framework :: AsyncIO"
]
```

File: setup.cfg [hunk 14]
```diff
@@ -1,55 +0,0 @@
-[metadata]
-name = smarthomeconnect
-version = 0.9.0
-url = https://github.com/mhthies/smarthomeconnect
-author = Michael Thies
-author_email = mail@mhthies.de
-project_urls =
-    Documentation = https://smarthomeconnect.readthedocs.org/
-    Source = https://github.com/mhthies/smarthomeconnect
-    Tracker = https://github.com/mhthies/smarthomeconnect/issues
```
[the remark this answers is not in the document]
Comment (jo47011) [reply setup.cfg:10]: 🙈 sorry, added again:
```bash
[project.urls]
Documentation = "https://smarthomeconnect.readthedocs.org/"
Source = "https://github.com/mhthies/smarthomeconnect"
Tracker = "https://github.com/mhthies/smarthomeconnect/issues"
```

File: setup.cfg [hunk 15]
```diff
@@ -1,55 +0,0 @@
...
-description = The Smart Home Connect home automation framework based on AsyncIO
-long_description = file: README.md
-long_description_content_type = text/markdown
-license = Apache License 2.0
-license_files = NOTICE, LICENSE
```
[the remark this answers is not in the document]
Comment (jo47011) [reply setup.cfg:15]: fixed:
```bash
[project]
...
readme = { file = "README.md", content-type = "text/markdown" }

[tool.setuptools]
license-files = ["LICENSE", "NOTICE"]
```

```bash
unzip -p smarthomeconnect-0.9.0-py3-none-any.whl smarthomeconnect-0.9.0.dist-info/METADATA

Metadata-Version: 2.2
Name: smarthomeconnect
Version: 0.9.0
Summary: The Smart Home Connect home automation framework based on AsyncIO
Author-email: Michael Thies <mail@mhthies.de>
License: 
                                         Apache License
                                   Version 2.0, January 2004
                                http://www.apache.org/licenses/
        
           TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
```

File: pyproject.toml [hunk 16]
```diff
@@ -1,3 +1,137 @@
...
+    "D400",  # First line should end with a period
+    "D401",  # First line of docstring should be in imperative mood
+    "D402",  # First line should not be the function's signature
+    "D404",  # First word of the docstring should not be "This"
+    "D415",  # First line should end with a period, question mark, or exclamation point
+    "E226",  # Missing whitespace around arithmetic operator  # FIXME: enable again
+    "E241",  # Multiple spaces after comma  # FIXME: enable again
+    "I001",  # Import block is un-sorted or un-formatted  # FIXME: enable again
+]
+preview = true
```
[the remark this answers is not in the document]
Comment (jo47011) [reply pyproject.toml:106]: removed.

File: scripts/extract_dependencies.py (hunk 7 above)
Comment (jo47011) [reply scripts/extract_dependencies.py:25]: Ok, no problem => removed.

File: setup.cfg (hunk 13 above)
[the remark this answers is not in the document]
Comment (mhthies) [reply setup.cfg:25]: Thanks! :)

I think we can keep  `Programming Language :: Python :: 3.8` and the versions in between, as well to show Python version compatibility on PyPI.

**Review by mhthies — APPROVED** [review]
